borg/archive.py

```diff
@@ -917,3 +922,288 @@ def finish(self, save_space=False):
+                # Re-visit all of these items in the archive even when fast-forwarding to rebuild hardlink_masters
+                hardlink_masters[item[b'path']] = (item.get(b'chunks'), None)
+                continue
+            if not matcher.match(item[b'path']):
+                logger.info('x %s', remove_surrogates(item[b'path']))
+                continue
+            if resume_from:
+                # Fast forward to after the last processed file
+                if item[b'path'] == resume_from:
+                    logger.info('Fast-forwarded to %s', remove_surrogates(item[b'path']))
+                    resume_from = None
+                continue
+            if target_is_subset and stat.S_ISREG(item[b'mode']) and item.get(b'source') in hardlink_masters:
+                # master of this hard link is outside the target subset
+                chunks, new_source = hardlink_masters[item[b'source']]
+                if new_source is None:
+                    # First item to use this master, move the chunks
+                    item[b'chunks'] = chunks
+                    hardlink_masters[item[b'source']] = (None, item[b'path'])
+                    del item[b'source']
+                else:
+                    # Master was already moved, only update this item's source
+                    item[b'source'] = new_source
+            if not self.dry_run:
+                try:
+                    self.process_item(archive, target, item)
+                except self.Interrupted:
+                    if self.progress:
+                        target.stats.show_progress(final=True)
+                    raise
+        assert resume_from is None, 'Reached end of source archive while fast-forwarding.'
+        if self.progress:
+            target.stats.show_progress(final=True)
+
+    def process_item(self, archive, target, item):
+        if b'chunks' in item:
+            self.process_chunks(archive, target, item)
+            target.stats.nfiles += 1
+        target.add_item(item)
+        if self.list:
+            logger.info('I %s', remove_surrogates(item[b'path']))
+        if self.interrupt:
+            raise self.Interrupted
+
+    def process_chunks(self, archive, target, item):
+        if not self.recompress and not self.rechunkify:
+            for chunk in item[b'chunks']:
+                self.cache.chunk_incref(chunk[0], target.stats)
+            return
+        chunk_iterator = archive.pipeline.fetch_many([c[0] for c in item[b'chunks']])
+        if self.rechunkify:
+            file = ChunkIteratorFileWrapper(chunk_iterator)
+            chunk_iterator = target.chunker.chunkify(file)
+        new_chunks = []
+        if target.rewrite_partial_chunks:
+            # No incref, create_target_or_resume already did that
+            new_chunks = target.rewrite_partial_chunks
+            target.rewrite_partial_chunks = None
+            for chunk in new_chunks:
+                if chunk[1] > 4096:
```

Comment:
small chunks are not put in the seen_chunks cache because they make little difference for performance but would still need the ~100 bytes/chunk or so in that cache.